tempodb/blocklist/poller.go

```diff
@@ -151,44 +154,90 @@ func (p *Poller) Do(previous *List) (PerTenant, PerTenantCompacted, error) {
 		return nil, nil, err
 	}
 
-	blocklist := PerTenant{}
-	compactedBlocklist := PerTenantCompacted{}
+	var (
+		wg  = boundedwaitgroup.New(p.cfg.TenantPollConcurrency)
+		mtx = sync.Mutex{}
+
+		blocklist          = PerTenant{}
+		compactedBlocklist = PerTenantCompacted{}
 
-	consecutiveErrors := 0
+		finalErr       atomic.Error
+		tenantFailures atomic.Int32
+
+		tenantFailuresRemaining = p.cfg.TolerateTenantFailures
+	)
 
 	for _, tenantID := range tenants {
-		newBlockList, newCompactedBlockList, err := p.pollTenantAndCreateIndex(ctx, tenantID, previous)
-		if err != nil {
-			level.Error(p.logger).Log("msg", "failed to poll or create index for tenant", "tenant", tenantID, "err", err)
-			consecutiveErrors++
-			if consecutiveErrors > p.cfg.TolerateConsecutiveErrors {
-				level.Error(p.logger).Log("msg", "exiting polling loop early because too many errors", "errCount", consecutiveErrors)
-				return nil, nil, err
+		// Exit early if we have exceeded our tolerance for number of failing tenants.
+		if tenantFailures.Load() >= int32(tenantFailuresRemaining) {
+			if lastErr := finalErr.Load(); lastErr != nil {
+				level.Error(p.logger).Log("msg", "exiting polling loop early because too many errors")
+				return nil, nil, lastErr
```

Comment:
i wouldn't return here. just `break` if we've had too many tenant failures and allow the method to exit naturally due to the check on 237-239.

Reply:
Good call, cleaner with fewer places to reason about the return.